esphome/components/hpma115c0/hpma115c0.cpp

```diff
@@ -0,0 +1,466 @@
+      this->adjustment_coefficient_needs_refresh_ = false;
+    }
+  }
+#endif  // USE_NUMBER
+
+  // Call all listeners' callbacks for value updates
+  for (auto &listener : this->listeners_)
+    listener->on_new_values(pm_1_0, pm_2_5, pm_4_0, pm_10_0, aqi_2_5, aqi_10_0);
+}
+
+// Scale a float value
+float Hpma115C0PollingComponent::scale_value_(float value, float in_min, float in_max, float out_min, float out_max) {
+  if (abs(in_max - in_min) <= 0.0)
+    return NAN;
+  return (value - in_min) / (in_max - in_min) * (out_max - out_min) + out_min;
+}
+
+// Compute and update frame CRC
+bool Hpma115C0PollingComponent::update_frame_crc_(StandardFrameT &frame) {
+  uint16_t crc;
+
+  // Sanity check on lentgh
+  if (frame.length > MAX_FRAME_DATA_LENGTH) {
+    ESP_LOGE(TAG, "Invalid length %d for CMD=0x%d", frame.length, frame.command);
+    this->hpma_last_error_ = HPMA_ERROR_DATA_LENGTH_TOO_LONG;
+    return false;
+  }
+
+  crc = frame.type + frame.length + frame.command;
+  for (int data_index = 0; data_index < frame.length - 1; data_index++) {
+    crc += frame.data[data_index];
+  }
+  crc = (0x10000UL - crc) & 0xFF;
+  frame.data[frame.length - 1] = (uint8_t) crc;
+
+  this->hpma_last_error_ = HPMA_ERROR_SUCCESS;
+  return true;
+}
+
+// Check frame CRC return computed
+bool Hpma115C0PollingComponent::check_frame_crc_(const StandardFrameT &frame) {
+  uint16_t crc;
+
+  // Sanity check on lentgh
+  if (frame.length > MAX_FRAME_DATA_LENGTH) {
+    ESP_LOGE(TAG, "Invalid length %d in reply %d for CMD=%02X", frame.length, frame.type, frame.command);
+    this->hpma_last_error_ = HPMA_ERROR_DATA_LENGTH_TOO_LONG;
+    return false;
+  }
+
+  // Compute CRC
+  crc = frame.type + frame.length + frame.command;
+
+  for (int data_index = 0; data_index < frame.length - 1; data_index++) {
+    crc += frame.data[data_index];
+  }
+
+  crc = (0x10000UL - crc) & 0xFF;
+
+  // Received frame CRC is located on byte after the data
+  if (((uint8_t) crc) != frame.data[frame.length - 1]) {
+    ESP_LOGE(TAG, "Invalid crc %02X (%02X expected) for reply %02X to CMD=%02X, LEN=%02X", frame.data[frame.length - 1],
+             crc, frame.type, frame.command, frame.length);
+    this->hpma_last_error_ = HPMA_ERROR_CRC;
+  }
+
+  this->hpma_last_error_ = HPMA_ERROR_SUCCESS;
+  return true;
+}
+
+// Clear UART buffer(s)
+void Hpma115C0PollingComponent::reset_sensor_mode_() {
+  uint32_t clear_start = millis();
+  uint32_t clear_duration, byte_count = 0;
+
+  // Empty UART buffer before sending command
+  byte_count = 0;
+  while ((clear_duration = (millis() - clear_start)) < MAX_SETUP_TIME) {
+    this->read();
+    byte_count += 1;
+  }
+  ESP_LOGD(TAG, "Emptied %d bytes from UART in %d ms", byte_count, clear_duration);
+
+  // Send command to stop autosend (default mode at power on)
+  ESP_LOGI(TAG, "Attempting to Stop Autosend");
+  if (!this->stop_autosend_()) {
+    ESP_LOGW(TAG, "Failed to stop autosend.");
```

Comment:
Bit of a nitpick, but do not include unnecessary punctuation in log messages (I see this in a number of messages later on, too). See our [logging best practices](https://developers.esphome.io/architecture/logging/).
```suggestion
    ESP_LOGW(TAG, "Failed to stop autosend");
```
Also, I think quite a few of these messages should be either `verbose` or `very_verbose`.

Reply:
Cleaned-up logging and decrease verbosity (thanks for the link on logging best practices).